### DISCUSSION TITLE
Cant connect robotstudio virtual controller to rws on ubuntu vm

### DISCUSSION THREAD

**Discussion by kalepo** [ros-industrial/abb_robot_driver#54] [Q&A]
I can't seem to connect the virtual controller to ros launch rws....does anyone have any ideas?

The port for the virtual controller is listening and it is the same as the one set on the Ubuntu vm roslaunch but its still not working. 

IP is set correctly as well as white listed
![Screenshot issue](https://user-images.githubusercontent.com/103397171/174515999-5c71d14a-968f-4411-9978-2cc95ce9d669.png)
![issue 2](https://user-images.githubusercontent.com/103397171/174516012-751e79b7-a4d1-429f-bbd0-0cfa39fc29b7.png)
![issue 3](https://user-images.githubusercontent.com/103397171/174516027-92eef7de-95af-4f33-b6a2-931b63127bc8.png)

**Comment by gavanderhoorn** [comment]
You are likely using a NAT-ed connection in your VM.

The driver attempts to connect to several ports opened by the VC, but if they are not forwarded, it will not be able to do so. The fact the "ports are listening" does not change this.

This is identical to how you cannot reach arbitrary ports behind a "nat router".

Either configure the VM network to forward those ports, or change the VM to a bridged or host-only network.

**Reply by kalepo** [reply]
Hey @gavanderhoorn my connection is a bridged adapter, I changed it and it still didnt work unfortunately

**Reply by kalepo** [reply]
Any other ideas around this?

**Reply by gavanderhoorn** [reply]
You'll have to figure out what's going on with your network configuration.

Perhaps you have a firewall active (on the VM guest or even the host), or are speccing the wrong IPs.

RWS is on `tcp://:80`. I'd suggest to check you can reach it from outside your VM (with `netcat` or a similar tool).

From the error message this doesn't seem like a problem with the driver at this point.

**Comment by gavanderhoorn** [comment]
Oh and off-topic perhaps, but please be aware that performance in a VM is going to be very sub-optimal.

VBox is not known for its performance, and while the driver does not have any UI, scheduling inside your VM is going to influence the behaviour of the driver (and any controllers you load) significantly.

If possible, reverse your setup (ie: native Linux, Windows-in-VM).
